### PR TITLE
Serial: implementation of TX FIFO

### LINKED ISSUE #25: serial: implement TX FIFO
Our serial console implements only one FIFO (the RX one).

We need to implement the TX FIFO as well because otherwise it is hard to follow the specification. Not implementing this fifo has the following side effect - we are flooding the guest with interrupts. In non-FIFO mode, an interrupt will be generated when the THR gets empty (therefore, for each received byte). In FIFO mode the interrupt is sent when the TX FIFO gets completely empty (every FIFO_SIZE bytes).
A 16550 spec example [here](http://caro.su/msx/ocm_de1/16550.pdf). It is worth mentioning, the spec defines the FIFO_SIZE as 16 bytes for UART 16550(A). However, it can be for example 64 bytes as well, as it can be seen [here](https://elixir.bootlin.com/linux/v5.8.9/source/drivers/tty/serial/8250/8250_port.c#L179).

### PR DESCRIPTION
**Pull request by FedericoPonzi**
Fixes #25 
To gain better general performance, a TX fifo for the Serial register is
introduced. With the THR fifo, a THR empty interrupt will be sent every
FIFO_SIZE bytes.

If the `out` descriptor is full, the content of the fifo is throwed
away.

### CODE REVIEW

**Comment by andreeaflorescu** [comment]
I'd like to integrate this PR in vmm-reference so that we can also test it end to end before merging it.

File: src/serial.rs [hunk 1]
```diff
@@ -270,7 +274,8 @@ impl<T: Trigger, EV: SerialEvents, W: Write> Serial<T, EV, W> {
             modem_control: DEFAULT_MODEM_CONTROL,
             modem_status: DEFAULT_MODEM_STATUS,
             scratch: DEFAULT_SCRATCH,
-            in_buffer: VecDeque::new(),
+            in_buffer: Default::default(),
+            out_written: Default::default(),
```
[the remark this answers is not in the document]
Comment (FedericoPonzi) [reply src/serial.rs:278]: No particular reason, just so that there is less stuff to update in case one wanted to change type for these fields (say from VecDequeue to Vec). I'll use explicit defaults.

File: src/serial.rs [hunk 2]
```diff
@@ -619,15 +634,17 @@ mod tests {
...
 
         serial.write(IER_OFFSET, IER_THR_EMPTY_BIT).unwrap();
         assert_eq!(
             serial.interrupt_enable,
             IER_THR_EMPTY_BIT & IER_UART_VALID_BITS
         );
-        serial.write(DATA_OFFSET, b'a').unwrap();
-
+        for _ in 0..FIFO_SIZE {
+            serial.write(DATA_OFFSET, b'a').unwrap();
```
[the remark this answers is not in the document]
Comment (FedericoPonzi) [reply src/serial.rs:646]: I've tried to use `assert_eq!(intr_evt.read().unwrap(), 0);` but fails with `Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }`. I'm checking directly for `IIR_THR_EMPTY_BIT`.

File: src/serial.rs [hunk 3]
```diff
@@ -204,7 +207,8 @@ pub struct Serial<T: Trigger, EV: SerialEvents, W: Write> {
     // functionality in FIFO mode. Reading from RBR will return the oldest
     // unread byte from the RX FIFO.
     in_buffer: VecDeque<u8>,
-
+    // Used to keep track of the amount of bytes written to the out device.
```
[the remark this answers is not in the document]
Comment (FedericoPonzi) [reply src/serial.rs]: Done. Please feel free to propose a different wording :relaxed: thanks for all your comments Andreaa!

File: src/serial.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply src/serial.rs:649]: We can actually use that error to validate that no interrupt was sent. So essentially we would do:
```rust
serial.write(DATA_OFFSET, b'a').unwrap();
// match on the error on read to be EWOULDBLOCK
let expected = io::Error::from(io::ErrorKind::WouldBlock);
assert!(matches!(intr_evt.read().unwrap_err(), expected));
```

**Review by lauralt — COMMENTED** [review]
I'm not sure if we should update the README in any way. Looks okay as it is right now. What do you think?

**Review by andreeaflorescu — APPROVED** [review]

**Review by andreeaflorescu — APPROVED** [review]

**Comment by andreeaflorescu** [comment]
@FedericoPonzi I did a rebase on you branch to get rid of the merge commit so that we can merge the PR.

**Review by andreeaflorescu — APPROVED** [review]

**Comment by andreeaflorescu** [comment]
Thanks @FedericoPonzi for the PR again. As we discussed offline, we decided to close this and not implement the TX FIFO for now. The reason is that the current implementation only accounts for Linux drivers, and the vm-superio should be usable on other operating systems as well. Reducing the number of IRQs would require us to implement the baud rate & other registers (such as FCR), and this adds unneeded complexity as the serial console is primarily used in debugging scenarios.
